### PR TITLE
Supports synthetic pod inter-pod anti-affinity

### PR DESCRIPTION
**Pull request by dposada**
## Changes proposed in this PR

Allowing (optionally) synthetic pods to have inter-pod anti-affinity to a configurable pod label.

## Why are we making these changes?

In order to repel synthetic pods from tenured nodes, e.g. by running a "repeller" pod on nodes that have been up and running for a certain amount of time.

### CODE REVIEW

File: scheduler/src/cook/kubernetes/api.clj [hunk 1]
```diff
@@ -1120,12 +1121,34 @@
...
-    ; grace period of 30 seconds can cause real job pods to be deemed unschedulable and fail.
     (when (synthetic-pod? pod-name)
+      ; We want to allow synthetic pods to have a non-default (typically 0) termination grace period,
+      ; so that they can be deleted quickly to free up space on nodes for real job pods. The default
+      ; grace period of 30 seconds can cause real job pods to be deemed unschedulable and fail.
       (when-let [{:keys [synthetic-pod-termination-grace-period-seconds]} (config/kubernetes)]
-        (.setTerminationGracePeriodSeconds pod-spec synthetic-pod-termination-grace-period-seconds)))
+        (.setTerminationGracePeriodSeconds pod-spec synthetic-pod-termination-grace-period-seconds))
+
+      ; We want to allow synthetic pods to be repelled from certain nodes via inter-pod (anti-)affinity
```
Comment (scrosby) [inline scheduler/src/cook/kubernetes/api.clj:1131]: We need to say why we're doing this.

File: scheduler/src/cook/kubernetes/api.clj [hunk 2]
```diff
@@ -1120,12 +1121,34 @@
...
+      ; (https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#inter-pod-affinity-and-anti-affinity)
+      (let [{:keys [synthetic-pod-anti-affinity-pod-label-key
+                    synthetic-pod-anti-affinity-pod-label-value]}
+            (config/kubernetes)]
+        (when (and synthetic-pod-anti-affinity-pod-label-key
+                   synthetic-pod-anti-affinity-pod-label-value)
+          ; The synthetic pod pod-spec may or may not already have an affinity defined
+          ; (see pod-hostnames-to-avoid above), so we need to allow for either case
+          (let [affinity (or (.getAffinity pod-spec) (V1Affinity.))
```
Comment (scrosby) [inline scheduler/src/cook/kubernetes/api.clj:1140]: I'd rephrase this as: 
""" If the synthetic pod already has an affinity (see ....)  we add to it, otherwise, we create a new one """

File: scheduler/src/cook/kubernetes/api.clj (hunk 1 above)
Comment (dposada) [reply scheduler/src/cook/kubernetes/api.clj:1131]: Good call, done

File: scheduler/src/cook/kubernetes/api.clj (hunk 2 above)
Comment (dposada) [reply scheduler/src/cook/kubernetes/api.clj:1144]: Agreed, done

**Review by scrosby — APPROVED** [review]
